internal/util-cache/src/main/scala/sbt/internal/util/AdditionalFormats.scala

```diff
@@ -0,0 +1,57 @@
+package sbt.internal.util
+
+import sbt.internal.util.Types.:+:
+
+import sjsonnew.{ Builder, deserializationError, JsonFormat, Unbuilder }
+import sjsonnew.BasicJsonProtocol, BasicJsonProtocol.asSingleton
+
+import java.io.{ ByteArrayInputStream, ByteArrayOutputStream, InputStream, OutputStream }
+
+trait SetFormat { self: BasicJsonProtocol =>
+  implicit def SetFormat[T: JsonFormat]: JsonFormat[Set[T]] = project(_.toSeq, (_: Seq[T]).toSet)
+}
+
+trait HListFormat {
+  implicit def HConsFormat[H: JsonFormat, T <: HList: JsonFormat]: JsonFormat[H :+: T] =
+    new JsonFormat[H :+: T] {
+      override def read[J](jsOpt: Option[J], unbuilder: Unbuilder[J]): H :+: T =
+        jsOpt match {
+          case Some(js) =>
+            unbuilder.beginObject(js)
+            val h = unbuilder.readField[H]("h")
+            val t = unbuilder.readField[T]("t")
+            unbuilder.endObject()
+
+            HCons(h, t)
+
+          case None =>
+            deserializationError("Expect JValue but found None")
+        }
+
+      override def write[J](obj: H :+: T, builder: Builder[J]): Unit = {
+        builder.beginObject()
+        builder.addField("h", obj.head)
+        builder.addField("t", obj.tail)
+        builder.endObject()
+      }
+    }
+
+  implicit val HNilFormat: JsonFormat[HNil] = asSingleton(HNil)
+
+}
+
+trait StreamFormat { self: BasicJsonProtocol =>
```

Comment:
I see. I'd say turn it into an Array of Byte first in that particular case, rather than adding unbounded streaming support.